tests/test_pd_controller.cpp

```diff
@@ -0,0 +1,210 @@
+// order to force conversion & co ...)
+template <typename ValueType, typename... T>
+constexpr auto MakeArray(T &&...values) -> std::array<ValueType, sizeof...(T)> {
+  return std::array<ValueType, sizeof...(T)>{std::forward<T>(values)...};
+}
+
+// RAII mutator, use to temporary modify a value given a reference and a tmp
+// value
+//
+// WARNING: Dangling references & co ...
+template <typename ValueType>
+struct TemporaryMutate {
+  TemporaryMutate() = delete;
+
+  template <typename T>
+  TemporaryMutate(ValueType &data, T &&tmp_value)
+      : data_(data), before_(data_) {
+    data_ = std::forward<T>(tmp_value);
+  }
+
+  virtual ~TemporaryMutate() { data_ = before_; }
+
+  constexpr auto GetOldValue() const -> const ValueType & { return before_; }
+
+ private:
+  ValueType &data_;
+  ValueType before_;
+};
+
+TEST(PdControllerTest, Ctor) {
+  EXPECT_NO_THROW({ const auto pd_ctrl = PDController(); });
+}
+
+struct Gains {
+  using IdxType = Eigen::VectorXd::Index;
+
+  Eigen::VectorXd p;
+  Eigen::VectorXd d;
+
+  static auto Random(IdxType p_size,
+                     std::optional<IdxType> d_size = std::nullopt) -> Gains {
+    return Gains{
+        .p = Eigen::VectorXd::Random(p_size),
+        .d = Eigen::VectorXd::Random(d_size.value_or(p_size)),
+    };
+  }
+};
+
+TEST(PdControllerTest, SetGains) {
+  auto pd_ctrl = PDController();
+
+  {
+    // FIXME: Is it valid ?
+    SCOPED_TRACE("Size 0");
+    EXPECT_NO_THROW(
+        { pd_ctrl.set_gains(Eigen::VectorXd{}, Eigen::VectorXd{}); });
+  }
+
+  {
+    SCOPED_TRACE("Same Sizes");
+    for (const auto both_size : MakeArray<Gains::IdxType>(1, 3, 4, 50, 1000)) {
+      auto requested_gains = Gains::Random(both_size);
+      EXPECT_NO_THROW(
+          { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+      // FIXME: How to test the validity ??
+
+      // Test for special double values acceptance
+      for (const auto special_value : {
+               std::numeric_limits<double>::infinity(),
+               std::numeric_limits<double>::quiet_NaN(),
+               std::numeric_limits<double>::signaling_NaN(),
+           }) {
+        {
+          auto _ = TemporaryMutate{requested_gains.p(0), special_value};
+          EXPECT_NO_THROW(
+              { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+        }
+
+        {
+          auto _ = TemporaryMutate{requested_gains.d(0), special_value};
+          EXPECT_NO_THROW(
+              { pd_ctrl.set_gains(requested_gains.p, requested_gains.d); });
+        }
+      }
+    }
+  }
+
+  {
+    SCOPED_TRACE("Different Sizes");
+    using SizeList = std::array<Gains::IdxType, 2>;
+    for (const auto [p_size, d_size] : MakeArray<SizeList>(
+             SizeList{1, 2}, SizeList{4, 3}, SizeList{50, 1000})) {
+      const auto requested_gains = Gains::Random(p_size, d_size);
+      // FIXME: I guess it should fail but set_gains doesn't provide any
+      // feedback on failure
```

Comment:
@MaximilienNaveau  could you elaborate if robot can have different number of `v` and `q`?

Reply:
Typically this behavior happens when you have some weird joints.
I believe we are tackling robots where each motor apply only 1 torque on one rotary joint.
We could use pinnochio for a generalization of this code. But I believe this approximation is fine for now.